lib/straight/gateway.rb

```diff
@@ -90,32 +89,33 @@ def new_order(args)
         if args[:amount].nil? || !args[:amount].kind_of?(Numeric) || args[:amount] <= 0
           raise OrderAmountInvalid, "amount cannot be nil and should be more than 0"
         end
-        # Setting default values
-        args[:currency]         ||= default_currency
-        args[:btc_denomination] ||= :satoshi
 
-        amount = args[:amount_from_exchange_rate] = amount_from_exchange_rate(
+        # Build order with default values
+        order = Kernel.const_get(order_class).new
+        order.gateway = self
+        order.keychain_id = args[:keychain_id]
+        order.block_height_created_at = fetch_latest_block_height rescue nil
+
+         args[:currency] ||= default_currency
+
+        args[:amount_from_exchange_rate] = amount_from_exchange_rate(
           args[:amount],
-          currency:         args[:currency],
-          btc_denomination: args[:btc_denomination]
+          currency: args[:currency],
+          btc_denomination: args[:btc_denomination] || :satoshi,
         )
```

Comment:
Don't really understand your 1) point. Do you consider bad that `args` changed or that it didn't?
Before, the hash was being (possibly) changed for the `:currency`, `:btc_denomination` and `:amount_from_exchange_rate` keys. Now the only key being changed is `:currency`, and only because is being used in various places in the function. We could actually also remove the `args[:amount_from_exchange_rate]` intermediate assignment and assign it directly to `order.amount`.

For 2), the `args[:btc_denomination]` param is only being used for the `amount_from_exchange_rate` function and thus I didn't consider it necessary to change or assign it to the hash attribute. `args[:currency]` is being used for various calculations and that's why I used  `args[:currency] ||= default_currency`.

I have no problem going back to the previous version, although personally I find this version easier to read and elegant enough, even if only for the "cosmetic" changes of grouping the steps logically like I described in the commit message.